### PR TITLE
8334223: Make Arena MEMFLAGs immutable

### PR DESCRIPTION
**Pull request by tstuefe**
Arenas carry NMT flags. 

An arena should never change that flag. But it does: Arenas (as ResourceAreas), used by CompilerThread, are accounted toward mtCompiler. But since the RA is already created in the parent class constructor (as mtThread), we then have to awkwardly change the flag of an already existing RA in the CompilerThread constructor.

As a prerequisite for future NMT work I would like Arena MEMFLAGS to be immutable.

The patch does that:
- we hand in MEMFLAGS to the Thread constructor now (defaults to mtThread)
- CompilerThread hands in mtCompiler, all other threads rely on the default
- on creation, both ResourceArea and HandleArea are now accounted toward the flag handed in
- that allows us to make Arena::flags const, and to remove ResourceArea::bias_to which changed the flag in-flight for the arena
- it also allows us to make Arena::flags private

Other, unrelated cleanups:
- Made Arena::_size_in_bytes and Arena::_tag private
- Merged both Arena constructors into one by specifying a default value of `Chunk::init_size` for `init_size` argument. That makes it equivalent to the old `Arena(flag, tag)` constructor
- removed `JavaThread::JavaThread(bool)`. That constructor was used when creating threads that are getting attached. There was only a single use for that constructor, and I replaced it with functionally equivalent code.

Tests:

I manually verified that the NMT numbers printed don't change.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334223](https://bugs.openjdk.org/browse/JDK-8334223): Make Arena MEMFLAGs immutable (**Enhancement** - P4)

### Reviewers
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**) ⚠️ Review applies to [78238a42](https://git.openjdk.org/jdk/pull/19693/files/78238a428ef48f31d1b6ad92c9e578862ff50875)
 * [Afshin Zafari](https://openjdk.org/census#azafari) (@afshin-zafari - Committer)
 * [Gerard Ziemski](https://openjdk.org/census#gziemski) (@gerard-ziemski - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19693/head:pull/19693` \
`$ git checkout pull/19693`

Update a local copy of the PR: \
`$ git checkout pull/19693` \
`$ git pull https://git.openjdk.org/jdk.git pull/19693/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19693`

View PR using the GUI difftool: \
`$ git pr show -t 19693`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19693.diff">https://git.openjdk.org/jdk/pull/19693.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19693#issuecomment-2169142821)

### CODE REVIEW

**Review by jdksjolen — APPROVED** [review]
Hi Thomas,

This looks like a very good cleanup to me, thank you.

All the best,
Johan

**Comment by tstuefe** [comment]
> This looks like a very good cleanup to me, thank you.

Thanks @jdksjolen !

I would like to have someone from the runtime team okay the changes to the thread constructors. Maybe @dholmes-ora ?

File: src/hotspot/share/runtime/javaThread.hpp [hunk 1]
```diff
@@ -1136,6 +1135,7 @@ class JavaThread: public Thread {
   // clearing/querying jni attach status
   bool is_attaching_via_jni() const { return _jni_attach_state == _attaching_via_jni; }
   bool has_attached_via_jni() const { return is_attaching_via_jni() || _jni_attach_state == _attached_via_jni; }
+  inline void set_is_attaching_via_jni();
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/javaThread.hpp:1138]: So the only thing I don't like about this aspect is that this has to be called at construction time, so having it be a stand-alone function call invites misuse. I'm tempted to add a factory method:
```
JavaThread* JavaThread::new_attaching_thread() {
  JavaThread* jt = new JavaThread();
  jt>set_is_attaching_via_jni();
}
```
and make `set_is_attaching_via_jni()` private.

What do you think?

File: src/hotspot/share/runtime/javaThread.inline.hpp [hunk 2]
```diff
@@ -196,6 +196,11 @@ void JavaThread::enter_critical() {
   _jni_active_critical++;
 }
 
+inline void JavaThread::set_is_attaching_via_jni() {
+  _jni_attach_state = _attaching_via_jni;
+  OrderAccess::fence();
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/javaThread.inline.hpp:201]: No need for a fence here as this should be set before the thread has been "published" and is visible to anyone else. In contrast `set_done_attaching_via_jni` is set after the thread has been published.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]

File: src/hotspot/share/runtime/javaThread.hpp (hunk 1 above)
Comment (tstuefe) [reply src/hotspot/share/runtime/javaThread.hpp:1138]: Yes, I like that better, too. Okay, I'll do that.

File: src/hotspot/share/runtime/javaThread.inline.hpp (hunk 2 above)
Comment (tstuefe) [reply src/hotspot/share/runtime/javaThread.inline.hpp:201]: I had the vague notion of not wanting to tie `set_done_attaching_via_jni` to an exact point in time. But that was not well thought out anyway. With a factory method, this problem disappears.

**Comment by tstuefe** [comment]
@dholmes-ora okay, I worked in your feedback. Thank you!

**Review by dholmes-ora — COMMENTED** [review]
Thanks for update - thread changes look good.

**Comment by tstuefe** [comment]
@gerard-ziemski or @afshin-zafari could you give this a glance, please? Its a cleanup related to NMT. Thank you!

**Comment by afshin-zafari** [comment]
Thanks, I could only found some missing Copyright updates:
- `compilerThread.cpp`
- `resourceArea.cpp` and its `hpp`
- `jni.cpp`, maybe for RedHat also?
- `handles.hpp`
- `thread.hpp`

**Comment by tstuefe** [comment]
> Thanks, I could only found some missing Copyright updates:
> 
> * `compilerThread.cpp`
> * `resourceArea.cpp` and its `hpp`
> * `jni.cpp`, maybe for RedHat also?
> * `handles.hpp`
> * `thread.hpp`

Many thanks, @afshin-zafari. Fixed copyrights.

**Comment by tstuefe** [comment]
@afshin-zafari if you are happy, mind approving? Thanks!

**Review by afshin-zafari — APPROVED** [review]
Usually when there are other pending reviews, I hold approving a PR.

**Review by gerard-ziemski — APPROVED** [review]
LGTM, thanks.

**Comment by tstuefe** [comment]
Thanks all!


/integrate

**Pull request closed without merging**